### PR TITLE
Optimize query planning for simple IN clauses

### PR DESCRIPTION
**Pull request by charlesmccarthyirl**
Simple queries with very large in clauses take a long time for planning in presto. A simple query such as of `SELECT COUNT(*) from my_table where id IN (<1 million IDs>)` took roughly 10 minutes in planning. The commits in this PR try to better optimize for these type of queries. They still take ~30 seconds in planning even after this change, but at least it's an improvement.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java [hunk 1]
```diff
@@ -149,33 +149,34 @@ public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
         {
             PlanNode source = context.rewrite(node.getSource());
 
-            TupleDomain<Symbol> tupleDomain = fromPredicate(metadata, session, node.getPredicate(), types).getTupleDomain();
-
+            // fromPredicate can be a very expensive operation. We should only call it if it's actually necessary.
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java]: In extreme case, most operations can become expensive, so I doubt this comment is that useful. I appreciate thorough explanation you did in the commit message though.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 2]
```diff
@@ -689,6 +691,25 @@ protected ExtractionResult visitInPredicate(InPredicate node, Boolean complement
             InListExpression valueList = (InListExpression) node.getValueList();
             checkState(!valueList.getValues().isEmpty(), "InListExpression should never be empty");
 
+            Symbol symbol = Symbol.from(node.getValue());
+            Type type = checkedTypeLookup(symbol);
+
+            // Check if we can bypass all the processing. If all the values in the IN clause are literals,
+            // and the type is orderable, we can just directly construct the SortedRangeSet.
+            if (type.isOrderable() && valueList.getValues().stream().allMatch(v -> v instanceof Literal)) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: `s/v -> v instanceof Literal/Literal.class::isInstance`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 3]
```diff
@@ -689,6 +691,25 @@ protected ExtractionResult visitInPredicate(InPredicate node, Boolean complement
...
+                ArrayList<Range> ranges = new ArrayList<>(valueList.getValues().size());
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: `s/ArrayList<Range>/List<Range>`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 4]
```diff
@@ -689,6 +691,25 @@ protected ExtractionResult visitInPredicate(InPredicate node, Boolean complement
...
+                ConnectorSession connectorSession = this.session.toConnectorSession();
+                for (Expression value : valueList.getValues()) {
+                    Object obj = LiteralInterpreter.evaluate(metadata, connectorSession, value);
+                    ranges.add(Range.equal(type, obj));
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: Better to stream this
```
valueList.getValues().stream()
 .map(...)
 .map(obj -> Range.equal(type, obj))
 .collect(toImmutableList())
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 5]
```diff
@@ -689,6 +691,25 @@ protected ExtractionResult visitInPredicate(InPredicate node, Boolean complement
             InListExpression valueList = (InListExpression) node.getValueList();
             checkState(!valueList.getValues().isEmpty(), "InListExpression should never be empty");
 
+            Symbol symbol = Symbol.from(node.getValue());
+            Type type = checkedTypeLookup(symbol);
+
+            // Check if we can bypass all the processing. If all the values in the IN clause are literals,
+            // and the type is orderable, we can just directly construct the SortedRangeSet.
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: Remove " Check if we can bypass all the processing.", it's pretty obvious that you special-case here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 6]
```diff
@@ -16,6 +16,7 @@
 import com.facebook.presto.Session;
 import com.facebook.presto.metadata.Metadata;
 import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.spi.ConnectorSession;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: "Optimize query planning for simple IN clauses" cmt msg -- you could be more specific in the commit title that you improve `DomainTranslator`. "Optimize query planning" sounds much more general.

**Comment by kokosing** [comment]
Can you please add a test for that?  Once you address comments, please ping me and I will merge this.

**Comment by charlesmccarthyirl** [comment]
@kokosing - I was talking with @jessesleeping - he's going to take it over since he has other ideas on how to make this a bit more generalized

**Comment by zhaiyuyong** [comment]
This pr doesn't work for my in operation
https://github.com/prestodb/presto/issues/9476

**Pull request closed without merging**